experimental/dds/tree2/src/feature-libraries/sequence-field/rebase.ts

```diff
@@ -674,53 +655,79 @@ function getMovedMark<T>(
 	return undefined;
 }
 
-function handleCurrAttach<T>(
-	currMark: Attach<T>,
-	factory: MarkListFactory<T>,
-	lineageRequests: LineageRequest<T>[],
-	offset: number,
-	baseIntention: RevisionTag | undefined,
+function handleLineage<T>(
+	rebasedMark: Mark<T>,
+	lineageRecipients: Mark<T>[],
+	lineageEntries: IdRange[],
+	baseIntention: RevisionTag,
 ) {
-	const rebasedMark = cloneMark(currMark);
-
 	// If the changeset we are rebasing over has the same intention as an event in rebasedMark's lineage,
 	// we assume that the base changeset is the inverse of the changeset in the lineage, so we remove the lineage event.
 	// TODO: Handle cases where the base changeset is a composition of multiple revisions.
 	// TODO: Don't remove the lineage event in cases where the event isn't actually inverted by the base changeset,
 	// e.g., if the inverse of the lineage event is muted after rebasing.
-	if (baseIntention !== undefined) {
-		tryRemoveLineageEvent(rebasedMark, baseIntention);
+	tryRemoveLineageEvents(rebasedMark, baseIntention);
+
+	for (const entry of lineageEntries) {
+		addLineageEntry(rebasedMark, baseIntention, entry.id, entry.count, entry.count);
 	}
-	factory.pushContent(rebasedMark);
-	lineageRequests.push({ mark: rebasedMark, offset });
+
+	lineageRecipients.push(rebasedMark);
 }
 
-interface LineageRequest<T> {
-	mark: Attach<T>;
-	offset: number;
+function addLineageToRecipients(
+	recipients: Mark<unknown>[],
+	revision: RevisionTag,
+	id: ChangesetLocalId,
+	count: number,
+) {
+	for (const mark of recipients) {
+		addLineageEntry(mark, revision, id, count, 0);
+	}
 }
 
-function updateLineage<T>(requests: LineageRequest<T>[], revision: RevisionTag) {
-	for (const request of requests) {
-		const mark = request.mark;
-		if (mark.lineage === undefined) {
-			mark.lineage = [];
-		}
+function addLineageEntry(
+	mark: Mark<unknown>,
+	revision: RevisionTag,
+	id: ChangesetLocalId,
+	count: number,
+	offset: number,
+) {
+	if (mark.lineage === undefined) {
+		mark.lineage = [];
+	}
 
-		mark.lineage.push({ revision, offset: request.offset });
+	if (mark.lineage.length > 0) {
+		const lastEntry = mark.lineage[mark.lineage.length - 1];
+		if (lastEntry.revision === revision && (lastEntry.id as number) + lastEntry.count === id) {
+			if (lastEntry.offset === lastEntry.count) {
+				mark.lineage[mark.lineage.length - 1] = {
+					...lastEntry,
+					count: lastEntry.count + count,
+					offset: lastEntry.offset + offset,
+				};
+				return;
+			} else if (lastEntry.offset === 0 && offset === 0) {
+				// Both entries are entirely after the mark
+				mark.lineage[mark.lineage.length - 1] = {
+					...lastEntry,
+					count: lastEntry.count + count,
+				};
+			}
+		}
```

Comment:
Why can we only extend the mark if both entries lie fully either before or after the mark?
For example, why can we not convert `[{ count: 2, offset: 1 }, { count: 2, offset: 0}]` to `{ count: 4, offset: 1 }`?
Does that case just never comes up?

Reply:
This code does merge the entries if the mark is after the first one but in the middle of the second. I'll update it to also merge if the mark is in the middle of the first one but before the second.